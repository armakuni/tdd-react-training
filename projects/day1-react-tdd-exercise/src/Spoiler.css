.spoiler {
  border: solid 3px #069;
}

.spoiler__header {
  background-color:#069;
  padding: 0.3em;
}

.spoiler__title {
  font-weight: bold;
  color: #eee;
  text-align: left;
}

.spoiler__button {
  float: right;
}

.spoiler__content {
  padding: 0.3em;
  background-color: #def;
  text-align: left;
}
