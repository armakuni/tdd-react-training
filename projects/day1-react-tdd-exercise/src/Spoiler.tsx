import React, { ReactElement, useState } from 'react';
import './Spoiler.css';

interface SpoilerProps {
  title: string
  content: string
}

const SpoilerState = {
  HIDDEN: 'hidden',
  SHOWING: 'showing'
}

const ButtonText = {
  SHOW: 'Show',
  HIDE: 'Hide'
}

export default function Spoiler({ title, content }: SpoilerProps): ReactElement {
  const [spoilerState, setSpoilerState] = useState(SpoilerState.HIDDEN);
  let buttonText : string;
  
  const isSpoilerShowing = () : boolean => {
    return spoilerState === SpoilerState.SHOWING 
  }

  const clickHandler = () => {
    setSpoilerState(isSpoilerShowing() ? SpoilerState.HIDDEN : SpoilerState.SHOWING);
  }

  buttonText = isSpoilerShowing() ? ButtonText.HIDE : ButtonText.SHOW;

  return (
    <>
      <button onClick={clickHandler}>{buttonText}</button>
      <h2>
        {title}
      </h2>
        {isSpoilerShowing() ? <div className='spoiler__content'>{content}</div> : '' }
    </>
  )
}
