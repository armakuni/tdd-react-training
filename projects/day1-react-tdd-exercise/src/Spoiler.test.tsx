import {
  fireEvent, render, RenderResult, screen,
} from '@testing-library/react';
import exp from 'constants';
import Spoiler from './Spoiler';

describe('<Spoiler />', () => {
  let container: HTMLElement;
  let screen: RenderResult;

  beforeEach(() => {
    screen = render(<Spoiler title="Example title" content="Secret message" />);
    container = screen.container;
  });

  it('always displays the text passed in the "title" prop', () => {
    expect(container.querySelector('h2')?.textContent).toBe('Example title');

    container = render(<Spoiler title="Different title" content="Secret message" />).container;
    expect(container.querySelector('h2')?.textContent).toBe('Different title');
  });

  it('has a "Show" button by default', () => {
    expect(container.querySelector('button')?.textContent).toBe('Show');
  });

  it('contains the spoiler content passed in the "content" prop', () => {
    container = render(<Spoiler title="Different title" content="Different secret message" />).container;
    fireEvent.click(container.getElementsByTagName('button')[0]);

    expect(container.querySelector('.spoiler__content')?.textContent).toBe('Different secret message')
  });

  it('should not show the spoiler content by default', () => {
    expect(screen.queryByText('Secret message')).not.toBeInTheDocument();
  });

  it('reveals the spoiler content is when the "Show" button is clicked', () => {
    fireEvent.click(container.getElementsByTagName('button')[0]);

    expect(screen.queryByText('Secret message')).toBeInTheDocument();
  });

  it('hides the spoiler content is when the "Hide" button is clicked', () => {
    
    const element = container.getElementsByTagName('button')[0];
    fireEvent.click(element);
    // given the spoiler is already showing
    expect(element.textContent).toBe('Hide');

    // when i click hide - 
    fireEvent.click(element);

    // then spoiler should be shown
    expect(screen.queryByText('Secret message')).not.toBeInTheDocument();
  });

});
