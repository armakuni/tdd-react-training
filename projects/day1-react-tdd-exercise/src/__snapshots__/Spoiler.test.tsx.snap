// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Spoiler /> Spoiler Component matches the spoiler-showing snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h1
          role="title"
        >
          Example title
        </h1>
        <button>
          Show
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h1
        role="title"
      >
        Example title
      </h1>
      <button>
        Show
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Spoiler /> Spoiler Component matches the spoiler-showing snapshot 2`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h1
          role="title"
        >
          Example title
        </h1>
        <p
          role="content"
        >
          Test content
        </p>
        <button>
          Hide
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h1
        role="title"
      >
        Example title
      </h1>
      <p
        role="content"
      >
        Test content
      </p>
      <button>
        Hide
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
